src/core/regimes.rkt

```diff
@@ -42,52 +42,37 @@
-  (define critexpr (critical-subexpression (*start-prog*)))
+  (define critexprs (all-critical-subexpressions (*start-prog*)))
   (define vars (program-variables (alt-program (car alts))))
 
-  (if critexpr
-      (remove-duplicates (cons critexpr vars))
+  (if critexprs
+      (remove-duplicates (append critexprs vars))
       vars))
 
-(define (critical-subexpression prog)
-  (define (loc-children loc subexpr)
-    (map (compose (curry append loc)
-		  list)
-	 (range 1 (length subexpr))))
-  (define (all-equal? items)
-    (if (< (length items) 2) #t
-	(and (equal? (car items) (cadr items)) (all-equal? (cdr items)))))
-  (define (critical-child expr)
-    (let ([var-locs
-	   (let get-vars ([subexpr expr]
-			  [cur-loc '()])
-	     (cond [(list? subexpr)
-		    (append-map get-vars (cdr subexpr)
-				(loc-children cur-loc subexpr))]
-		   [(constant? subexpr)
-		    '()]
-		   [(variable? subexpr)
-		    (list (cons subexpr cur-loc))]))])
-      (cond [(null? var-locs) #f]
-            [(all-equal? (map car var-locs))
-             (caar var-locs)]
-            [#t
-             (let get-subexpr ([subexpr expr] [vlocs var-locs])
-               (cond [(all-equal? (map cadr vlocs))
-                      (get-subexpr (if (= 1 (cadar vlocs)) (cadr subexpr) (caddr subexpr))
-                                   (for/list ([vloc vlocs])
-                                     (cons (car vloc) (cddr vloc))))]
-                     [#t subexpr]))])))
-  (let* ([locs (localize-error prog)])
-    (if (null? locs)
-        #f
-        (let* ([candidate-expr (critical-child (location-get (car locs) prog))]
-               [candidate-prog `(lambda ,(program-variables (*start-prog*)) ,candidate-expr)])
-          (if (and candidate-expr
-                   (for/or ([(pt ex) (in-pcontext (*pcontext*))])
-                     (nan? ((eval-prog candidate-prog 'fl) pt))))
-              #f
-              candidate-expr)))))
+(define (critical-subexpression? expr loc)
+  (let* ([crit-vars (free-variables (location-get loc expr))]
+         [replaced-expr (replace-expression expr (location-get loc expr) 1)]
+         [non-crit-vars (free-variables (location-get (list 2) replaced-expr))])
+    (set-disjoint? crit-vars non-crit-vars)))
+
+(define (all-critical-subexpressions prog)
+  (define (subexprs-in-expr expr loc curr-num)
+    (with-handlers ([exn:fail? (λ (e) null)])
+      (let* ([new-loc (append loc (list curr-num))]
+             [curr-subexpr (location-get new-loc expr)]
+             [subexpr-locs (append (subexprs-in-expr expr loc (add1 curr-num))
+                                   (subexprs-in-expr expr new-loc 1))])
+        (cond
+          [(null? (free-variables curr-subexpr))
+             subexpr-locs]
+          [#t
+           (if (critical-subexpression? expr new-loc)
+               (append (list new-loc) subexpr-locs)
+               subexpr-locs)]))))
+  (for/list ([loc (if (member (car prog) '(lambda λ))
```

Comment:
Often it is better to pick an input format than to handle multiple formats. That makes it easier to reason about the rest of the code, which reduces bugs. I'd rather require this code to take `lambda` terms, which I think it always does, than to have this conditional.